File: CardPeak.UI.ClientApp/src/services/api/dashboardController.ts
import axios from 'axios';

const API = {
    GET_DASHBOARD: '/dashboard',
    REFRESH_DASHBOARD: '/dashboard/refresh',
    GET_YEARS: '/dashboard/years'
};

export function getDashboard(
    successCallback: (data: CardPeak.Models.DashboardModel) => void,
    errorCallback?: (error: string) => void
) {
    axios
        .get(API.GET_DASHBOARD)
        .then(r => {
            successCallback(r.data as CardPeak.Models.DashboardModel);
        })
        .catch(reason => {
            if (errorCallback) {
                errorCallback(reason.message);
            }
        });
}

export function refreshDashboard(
    year: number,
    month: number,
    successCallback: (data: CardPeak.Models.DashboardModel) => void,
    errorCallback?: (error: string) => void
) {
    axios
        .get(API.REFRESH_DASHBOARD, {
            params: {
                year,
                month
            }
        })
        .then(r => {
            if (successCallback) {
                successCallback(
                    r.data as CardPeak.Models.DashboardModel
                );
            }
        })
        .catch(reason => {
            if (errorCallback) {
                errorCallback(reason.message);
            }
        });
}

export function getAvailableYears(
    successCallback: (
        data: CardPeak.Entities.ApprovalMetric<number>[]
    ) => void
) {
    axios.get(API.GET_YEARS, {}).then(r => {
        if (successCallback) {
            successCallback(
                r.data as CardPeak.Entities.ApprovalMetric<number>[]
            );
        }
    });
}

File: CardPeak.WebAPI/Controllers/DashboardController.cs
﻿using CardPeak.Core.Service;
using CardPeak.Service;
using System;
using System.Web.Http;

namespace CardPeak.WebAPI.Controllers
{
	[RoutePrefix("api/dashboard")]
	public class DashboardController : ApiController
	{
		private IDashboardService DashboardService;

		public DashboardController()
		{
			this.DashboardService = new DashboardService(new Repository.EF.CardPeakDbContext());
		}

		public IHttpActionResult Get()
		{
			return this.RefreshDashboard(DateTime.Now.Year, null);
		}

		[HttpGet]
		[Route("refresh")]
		public IHttpActionResult RefreshDashboard([FromUri]int? year = null, [FromUri]int? month = null)
		{
			var result = this.DashboardService.GetDashboard(year, month);
			if (result == null)
			{
				return this.NotFound();
			}
			return this.Ok(result);
		}

		[HttpGet]
		[Route("years")]
		public IHttpActionResult GetAvailableYears()
		{
			var result = this.DashboardService.GetAvailableYears();
			if (result == null)
			{
				return this.NotFound();
			}
			return this.Ok(result);
		}
	}
}
